core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1933,63 +1933,78 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
-      // We need to use `attempt` because `interruption` may already be completed.
-      val signalInterruption: F2[Unit] = interrupt.complete(()).void
-
-      def go(s: Stream[F2, O2], guard: Semaphore[F2]): Pull[F2, Nothing, Unit] =
-        Pull.eval(guard.acquire) >> s.pull.uncons.flatMap {
-          case Some((hd, tl)) =>
-            val send = resultChan.send(Stream.chunk(hd).onFinalize(guard.release))
-            Pull.eval(send) >> go(tl, guard)
-          case None => Pull.done
-        }
+  )(implicit F: Concurrent[F2]): Stream[F2, O2] =
+    Stream.force {
+      // `State` describes the state of an upstream stream (`this` and `that` are both upstream streams)
+      // None            : the stream is has not yet terminated
+      // Some(Left(t))   : the stream terminated with an error
+      // Some(Right(())) : the stream terminated successfully
+      type State = Option[Either[Throwable, Unit]]
+      // `bothStates` keeps track of the state of `this` and `that` stream so we can terminate the
+      // downstream when both upstreams terminate.
+      SignallingRef.of[F2, (State, State)]((None, None)).flatMap { bothStates =>
+        // `output` is used to send chunks from upstreams to downstream. It sends streams, not chunks,
+        // as each chunk has a finalizer associated with it.
+        Channel.synchronous[F2, Stream[F2, O2]].flatMap { output =>
+          // `stopDef` is used to interrupt the streams in the case that an upstream errors, or
+          // in the case that the downstream terminates.
+          Deferred[F2, Unit].map { stopDef =>
+            val signalStop: F2[Unit] = stopDef.complete(()).void
+            val stop: F2[Either[Throwable, Unit]] = stopDef.get.as(Right(()))
+            def complete(result: Either[Throwable, Unit]): F2[Unit] =
+              bothStates.update {
+                case (None, None)  => (Some(result), None)
+                case (other, None) => (other, Some(result))
+                case _             => sys.error("impossible")
+              }
+            val bothStopped: PartialFunction[(State, State), Either[Throwable, Unit]] = {
+              case (Some(r1), Some(r2)) => CompositeFailure.fromResults(r1, r2)
+            }
+            def run(s: Stream[F2, O2]): F2[Unit] =
+              // `guard` ensures we do not pull another chunk until the previous one has been consumed downstream.
+              Semaphore[F2](1).flatMap { guard =>
+                Stream
+                  .exec(guard.acquire)
+                  .append(s)
+                  .chunks
+                  .evalMap(chk =>
+                    output.send(
+                      Stream.chunk(chk).onFinalize(guard.release)
+                    )
+                  )
+                  .evalTap(_ => guard.acquire)
+                  // Stop when the other upstream has errored or the downstream has completed.
+                  .interruptWhen(stop)
```

Comment:
Would it be worth pointing out that, because the initial `guard.acquire` is at the  beginning of the stream, the `stop` can interrupt it as well?